### PR TITLE
importing an excel into a postgresql database directly

### PR DESCRIPTION
**Pull request by jpbeltrami**
issue #4

### CODE REVIEW

File: excel.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import xlrd
+from collections import OrderedDict
+import simplejson as json
+import psycopg2
+import urllib.parse
+
+# set up the connection. Extract the database info from the URI and connect with psycopg2
+def connect():
+    try:
+        result = urllib.parse.urlparse("postgres://4cebab44:4j602JWnaP3Yaw@db.nuodata.io:5432/4cebab44")
```
Comment (rngadam) [inline excel.py]: NEVER put authentication information in a git history. 

Easiest is to put it into a separate user-created JSON file and use built-in JSON parsing support.

```
rngadam@rngadam-mac ~/c/s/nuodata-example> cat config.json 
{
    "username": "USERNAME",
    "password": "PASSWORD"
}
rngadam@rngadam-mac ~/c/s/nuodata-example> python
>>> import json
>>> c = json.load(open('config.json'))
>>> c['username']
u'USERNAME'
>>> c['password']
u'PASSWORD'
```

Make sure to put config.json in .gitignore and create a config.json.example.

File: excel.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        DB = psycopg2.connect(
+            database = database,
+            user = username,
+            password = password,
+            host = hostname
+        )
+        c = DB.cursor()
+        return DB, c
+    except:
+        print("<couldn't connect>")
```
Comment (rngadam) [inline excel.py]: You've just created a function connect() that silently fails if a failure happens.

Best practice is to capture and log the full error information and rethrow a custom exception to force the caller to handle it.

File: excel.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+# opens the excel file, extracts data row by row, returns it in JSON
+def tojson():
+    # Open the workbook.
+    wb = xlrd.open_workbook('practice.xlsx')
```
Comment (rngadam) [inline excel.py]: Much less reusable if you put plain constant strings within a function. make 'practice.xlsx' a function parameter.

File: excel.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        excel_list.append(products)
+    return excel_list
+
+# uses psycopg2 to create a query to update the DB.
+def post():
+    jdata = tojson()
+    DB, c = connect()
+    for item in jdata:
+        c.execute("INSERT INTO test (name, age) VALUES (%s, %s)", (item['name'], item['age']))
```
Comment (rngadam) [inline excel.py]: better error handling required, read this:

http://stackoverflow.com/questions/8497886/graceful-primary-key-error-handling-in-python-psycopg2

**Review by rngadam — CHANGES_REQUESTED** [review]
* Don't put sensitive values in commit history
* Apply defensive programming practices
* More descriptive commit messages, pull requests names, function names ("my changes", "jdevelop", "practice.xlsx") are all meaningless names with no value

**Comment by rngadam** [comment]
update the README.md to explain what this does, put references and give usage example

File: excel.py (hunk 3 above)
Comment (jpbeltrami) [reply excel.py]: fixed

File: excel.py (hunk 2 above)
Comment (jpbeltrami) [reply excel.py]: fixed

File: excel.py (hunk 4 above)
Comment (jpbeltrami) [reply excel.py]: fixed

File: excel.py (hunk 1 above)
Comment (jpbeltrami) [reply excel.py]: fixed

File: excel.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+#! python3
+# Reads an excel file and transforms to JSON.
+
+import xlrd
+from collections import OrderedDict
+import simplejson as json
+import psycopg2
+import urllib.parse
+
+#
```
Comment (rngadam) [inline excel.py]: empty comment?

File: excel.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+# set up the connection. Extract the database info from the URI and connect with psycopg2
+def connect():
+    # enter your database API information.. either with the uri or the credentials
+    try:
+        result = urllib.parse.urlparse("postgres://USERNAME:PASSWORD@db.nuodata.io:5432/4cebab44")
```
Comment (rngadam) [inline excel.py]: I think this already works with connect()

File: excel.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        DB = psycopg2.connect(
+            database = database,
+            user = username,
+            password = password,
+            host = hostname
+        )
+        c = DB.cursor()
+        return DB, c
+    except psycopg2.Error as e:
+        print(e.pgerror)
```
Comment (rngadam) [inline excel.py]: this does not address the original comment asking to rethrow a custom exception

File: excel.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+# opens the excel file, extracts data row by row, returns it in JSON
+def tojson(workbook):
+    # Check the file extension is correct
+    if workbook.endswith('.xlsx'):
```
Comment (rngadam) [inline excel.py]: useless/undesirable check

File: excel.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            products['age'] = row_values[1]
+
+            excel_list.append(products)
+        return excel_list
+    else:
+        print('The file extension is not supported or is in uppercase.')
+
+# uses psycopg2 to create a query to update the DB.
+def post():
+    jdata = tojson('practice.xlsx')
```
Comment (rngadam) [inline excel.py]: and why not make this a parameter to post() too?  as the database parameter? and then, to make your tool really useful, parse command-line parameter instead:

https://docs.python.org/2/library/argparse.html

File: README.md [hunk 10]
```diff
@@ -1,2 +1,9 @@
 # nuodata-example
 Example for Nuodata.io
+
+- tojson method found on: http://www.anthonydebarros.com/2013/02/05/get-json-from-excel-using-python-xlrd/
```
Comment (rngadam) [inline README.md]: this does not explain how to use your utility.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,34 @@
-# nuodata-example
-Example for Nuodata.io
+## Synopsis
+This file takes an excel file and sends it to the specified URI database. In this case it's made for sending updates to a database on nuodata.io.
+
+## Pre-reqs
+```
+brew install python3
+pip install psycopg2
```
Comment (rngadam) [inline README.md]: use a requirements.txt file:

https://pip.pypa.io/en/stable/user_guide/

File: README.md [hunk 12]
```diff
@@ -1,2 +1,34 @@
...
+pip install xlrd
+```
+
+## Running
+```
+$ python3 excel.py excel-file database-URI name of database
```
Comment (rngadam) [inline README.md]: uppercase for parameters 

```
+$ python3 excel.py excel-file $DATABASE_URI $DATABASE_NAME
```

File: README.md [hunk 13]
```diff
@@ -1,2 +1,34 @@
...
+For example:
+```
+postgres://[username:password@]host[:PORT]/username
+```
+
+## Motivation
+This was made as practice to a bigger project that sends the same type of data to an API.
+
+## Installation
+Download the excel.py file to any directory.
```
Comment (rngadam) [inline README.md]:
```
git clone repo
```

File: README.md [hunk 14]
```diff
@@ -1,2 +1,34 @@
...
+
+## Reference
+- tojson method found on: http://www.anthonydebarros.com/2013/02/05/get-json-from-excel-using-python-xlrd/
+
+- error handling seen on: http://stackoverflow.com/questions/8497886/graceful-primary-key-error-handling-in-python-psycopg2
+
+## License
```
Comment (rngadam) [inline README.md]: use a LICENSE.TXT instead

File: excel.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+import xlrd
+import sys
+import simplejson as json
+from collections import OrderedDict
+import psycopg2
+import argparse
+import urllib.parse
+
+# TODO update readme with what this does, and usage example.
```
Comment (rngadam) [inline excel.py]: TODO as Github Issues instead

File: excel.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    parser.add_argument('dbname', help='the name of the database')
+
+    args = parser.parse_args()
+
+    xcel = args.xcel
+    uri = args.uri
+    dbname = args.dbname
+
+    # connect to the database using connect() (see below)
+    DB, c= connect(uri)
```
Comment (rngadam) [inline excel.py]: spacing before =

File: excel.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    # get the list of rows from the excel file
+    excel_list = toJson(xcel)
+
+    # send the items in the list to the db
+    post(excel_list)
+
+    # close the connection to the cursor and then to the database
+    c.close()
+    DB.close()
+##############################################
```
Comment (rngadam) [inline excel.py]: 2 whitelines before start of new functions

**Review by rngadam — CHANGES_REQUESTED** [review]

File: excel.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+        excel_list.append(products)
+    return excel_list
+
+
+# uses psycopg2 to create a query to update the DB.
+def post(DB, c, tableName, excel_list):
+    try:
+        for item in excel_list:
+            query = "INSERT INTO %s (product_name, quantity) VALUES ('%s', %s)" % (tableName, item['product_name'], item['quantity'])
```
Comment (rngadam) [inline excel.py]: try to keep lines <80

File: requirements.txt [hunk 19]
```diff
@@ -0,0 +1,2 @@
+pip install psycopg2
```
Comment (rngadam) [inline requirements.txt]: shouldn't it just be the package?

File: README.md [hunk 20]
```diff
@@ -1,2 +1,31 @@
-# nuodata-example
-Example for Nuodata.io
+## Synopsis
+This file takes an excel file and sends it to the specified URI database. In this case it's made for sending updates to a database on nuodata.io.
+
+## Pre-reqs
+run the following in the command line:
+```
+pip install -r requirements.txt
```
Comment (rngadam) [inline README.md:7]: test it in a clean python environment using virtualenv.

http://docs.python-guide.org/en/latest/dev/virtualenvs/

File: excel.py (hunk 18 above)
Comment (jpbeltrami) [reply excel.py]: fixed

File: README.md (hunk 20 above)
Comment (jpbeltrami) [reply README.md:7]: done

File: requirements.txt (hunk 19 above)
Comment (jpbeltrami) [reply requirements.txt]: fixed

**Pull request merged**